File: solutions/cpp/say/1/say.cpp
#include "say.h"
#include <string>
#include <stdexcept>
#include <array>

using namespace std;

namespace say {

     const array<string,4> power_names = {"hundred","thousand", "million", "billion"};
    const array<string,8> tens = {"twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"};
    const array<string,10> teens = {"ten", "eleven", "twelve", "thireen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"};
    const array<string,9> ones = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

    string parse_ones(unsigned long number) {
        if(number<1 || number>9)
            return "";
        return ones.at(number-1); 
    } 
    
    string parse_tens(unsigned long number) {
        unsigned long d = number/10;
        unsigned long r = number%10;
        switch(d) {
            case 0: return parse_ones(r);
                    break;
            case 1: return teens.at(r);
                    break;
            default: 
                string s1 = tens.at(d-2);
                if(r == 0)
                    return s1;
                else
                    return s1 + string("-") +  parse_ones(r);                    
        }
    }
    
    string parse_hundreds(unsigned long number) {      
        unsigned long d = number/100;
        unsigned long r = number%100;
        if(!d) 
            return parse_tens(r);
        else 
            {
                string sr = (r!=0)?  string(" ") + parse_tens(r): string("");
                return parse_ones(d) + string(" ") + power_names.at(0) + sr;
            }
            
    }
    
    string in_english(unsigned long number) {
        string result; 
        if(!number)
            return string("zero");
        else if (number>= 1000ULL*1000ULL*1000ULL*1000ULL)
            throw domain_error("Unsuported input.");
        string pad_previous = "";
        for(size_t i = 0; i<power_names.size() &&  number>0; i++) {
            auto r = number % 1000ULL;
            string s = parse_hundreds(r);
            if(i && s!="")  
                s += " " + power_names.at(i);
            result = s + pad_previous + result; 
            pad_previous = (r!=0) ? " " : "";
            number = (number - r)/1000ULL;            
        }
        return result;            
    }
    

}  // namespace say

File: solutions/python/say/1/say.py
powers = ["hundred", "thousand", "million", "billion"]
tens = ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
teens = ["ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]
ones = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

def parse_ones(number):
    if number<1 or number>9:
        return ""
    else:
        return ones[number-1]

def parse_tens(number):
    d = number//10
    r = number % 10
    if d == 0:
        return parse_ones(r)
    elif d == 1:
        return teens[r]
    else:
        str = tens[d-2]
        if r!=0:
            str += "-" + parse_ones(r)
        return str

def parse_hundreds(number):
    d = number // 100
    r = number % 100
    if d==0:
        return parse_tens(number)
    else:
        sr = ""
        if r!=0:
            sr = " " + parse_tens(r)
        return parse_ones(d) + " " + powers[0] + sr

def say(number):
    if number==0:
        return "zero"
    elif number<0 or number>999999999999:
        raise ValueError("input out of range")
    else:
        result = ""
        pad_previous = ""
        for i in range(4):
            if number==0:
                break
            r = number%1000
            str = parse_hundreds(r)
            if i>0 and str!="":
                str += " " + powers[i] 
            result = str + pad_previous + result
            if r!=0:
                pad_previous = " "
            else:
                pad_previous = ""
            number = (number-r)//1000
        return result
                
            
